File: lib/net/packet/data_packet.ex
defmodule Net.Packet.DataPacket do
  use Protobuf, protoc_gen_elixir_version: "0.10.0", syntax: :proto3

  field :type, 1, type: :int32
  oneof :payload, 2

  field :login, 2, type: Net.Packet.Login, oneof: 2
  field :auth_response, 3, type: Net.Packet.AuthResponse, oneof: 2
  field :update, 4, type: Net.Packet.Update, oneof: 2
end

File: client/ts/proto/DataPacket.ts
import { Message, Field, OneOf } from "protobufjs";
import LoginRequest from "./requests/LoginRequest";
import UpdateRequest from "./requests/UpdateRequest";
import AuthResponse from "./responses/AuthResponse";
import DataRequest from "./requests/DataRequest";

export default class DataPacket extends Message<DataPacket> {
  @Field.d(1, "int32", "required")
  public type!: number;
  
  @OneOf.d("login", "authResponse", "update", "request")
  public payload!: LoginRequest | AuthResponse | UpdateRequest | DataRequest;

  @Field.d(2, LoginRequest)
  login?: LoginRequest;

  @Field.d(3, AuthResponse)
  authResponse?: AuthResponse;

  @Field.d(4, UpdateRequest)
  update?: UpdateRequest;

  @Field.d(5, DataRequest)
  request?: DataRequest
}